tests/Bus/MongoBatchRepositoryTest.php

```diff
@@ -0,0 +1,364 @@
+
+        $batch->options['then'] = [];
+        $this->assertFalse($batch->hasThenCallbacks());
+        $batch->options['then'] = [1];
+        $this->assertTrue($batch->hasThenCallbacks());
+
+        $this->assertFalse($batch->allowsFailures());
+        $batch->options['allowFailures'] = true;
+        $this->assertTrue($batch->allowsFailures());
+
+        $this->assertFalse($batch->hasFailures());
+        $batch->failedJobs = 1;
+        $this->assertTrue($batch->hasFailures());
+
+        $batch->options['catch'] = [];
+        $this->assertFalse($batch->hasCatchCallbacks());
+        $batch->options['catch'] = [1];
+        $this->assertTrue($batch->hasCatchCallbacks());
+
+        $this->assertFalse($batch->cancelled());
+        $batch->cancelledAt = now();
+        $this->assertTrue($batch->cancelled());
+
+        $this->assertIsString(json_encode($batch));
+    }
+
+    /** @see BusBatchTest:test_chain_can_be_added_to_batch: */
+    public function testChainCanBeAddedToBatch()
+    {
+        $queue = m::mock(Factory::class);
+
+        $batch = $this->createTestBatch($queue);
+
+        $chainHeadJob = new ChainHeadJob();
+
+        $secondJob = new SecondTestJob();
+
+        $thirdJob = new ThirdTestJob();
+
+        $queue->shouldReceive('connection')->once()
+            ->with('test-connection')
+            ->andReturn($connection = m::mock(stdClass::class));
+
+        $connection->shouldReceive('bulk')->once()->with(m::on(function ($args) use ($chainHeadJob, $secondJob, $thirdJob) {
+            return $args[0] === $chainHeadJob
+                && serialize($secondJob) === $args[0]->chained[0]
+                && serialize($thirdJob) === $args[0]->chained[1];
+        }), '', 'test-queue');
+
+        $batch = $batch->add([
+            [$chainHeadJob, $secondJob, $thirdJob],
+        ]);
+
+        $this->assertEquals(3, $batch->totalJobs);
+        $this->assertEquals(3, $batch->pendingJobs);
+        $this->assertSame('test-queue', $chainHeadJob->chainQueue);
+        $this->assertIsString($chainHeadJob->batchId);
+        $this->assertIsString($secondJob->batchId);
+        $this->assertIsString($thirdJob->batchId);
+        $this->assertInstanceOf(CarbonImmutable::class, $batch->createdAt);
+    }
+
+    /** @see BusBatchTest::createTestBatch() */
+    private function createTestBatch(Factory $queue, $allowFailures = false)
+    {
+        $connection = DB::connection('mongodb');
+        $this->assertInstanceOf(Connection::class, $connection);
+
+        $repository = new MongoBatchRepository(new BatchFactory($queue), $connection, 'job_batches');
+
+        $pendingBatch = (new PendingBatch(new Container(), collect()))
+            ->progress(function (Batch $batch) {
+                $_SERVER['__progress.batch'] = $batch;
+                $_SERVER['__progress.count']++;
+            })
+            ->then(function (Batch $batch) {
+                $_SERVER['__then.batch'] = $batch;
```

Comment:
We can't use `$this` because the closure is serialized.

Reply:
How would `$this` be used here even if it _was_ supported?

Offhand, is [laravel/serializable-closure](https://github.com/laravel/serializable-closure) used internally for serialization? AFAIK, the stock `serialize()` function doesn't support closures on its own.

Reply:
All the `$_SERVER['something']` could have been replaced by `$this->data['something']`. But that's not allowed by the package you mentioned.